### PR TITLE
Introduce XState for managing state in the app

### PR DESCRIPTION
**Pull request by fraserdrops**
This PR uses XState to replace Jotai as the state manager for Katachidraw. 

**Purpose**
The aim is to replicate the existing functionality as closely as possible, so that this could be merged if desired. Otherwise, this code could serve as a reference for an alternative XState implementation, or even just as a comparison with the jotai implementation which is interesting in itself.

**To do:**
- History is not implemented yet. It is possible with XState, but it's a tricky feature to get right.
- Typings are not fully completed (I'm new to Typescript, so any help with these is welcome)
- Image rotation is not added yet.
- Code is duplicated between ShapeMachine and ImageMachine. I am still learning the best approach when there is shared code between machines.
- I didn't use jotai/xstate because I'm not familiar with it. At this stage I don't see what jotai/xstate would add compared to using xstate/react, which comes with the selectors and inspector already. However, it would hopefully be easy to swap one out for the other.

### CODE REVIEW

File: src/components/Canvas.tsx [hunk 1]
```diff
@@ -73,8 +65,8 @@ export const Canvas: FC<Props> = ({
           height={height / zoom}
           opacity="0"
         />
-        {ShapesElement}
-        {DotsElement}
+        <Shapes shapes={shapes} />
+        <Dots dots={dots} />
```
Comment (fraserdrops) [inline src/components/Canvas.tsx:69]: I swapped to declaring the components inline so I could add props to them. Was there a reason they were declared as props on the Canvas component originally

File: src/components/Shape.tsx [hunk 2]
```diff
@@ -1,102 +1,92 @@
...
 const ShapePath: React.FC<{
-  shapeAtom: ShapeAtom;
-  shape: TShapePath;
-}> = ({ shapeAtom, shape }) => {
-  const [mode] = useAtom(modeAtom); // XXX this is very unfortunate (re-render all shapes)
-  const [, select] = useAtom(selectAtom);
-  const [, setPressingShape] = useAtom(setPressingShapeAtom);
-  const [, registerIsPointInShape] = useAtom(registerIsPointInShapeAtom);
+  service: any;
+}> = React.memo(({ service }) => {
```
Comment (fraserdrops) [inline src/components/Shape.tsx:13]: The React.memo works because service is a stable ref

File: src/components/Toolbar.tsx [hunk 3]
```diff
@@ -1,54 +1,101 @@
...
+
+  {
+    id: "color",
+    component: PaletteIcon,
+    eventType: "SELECT_COLOR",
+    active: stateValue.mode === "color",
+  },
+  { id: "save", component: SaveIcon, eventType: "SELECT_SAVE" },
+  { id: "image", component: ImageIcon, eventType: "UPLOAD_IMAGE" },
+];
```
Comment (fraserdrops) [inline src/components/Toolbar.tsx:78]: This array seems a bit ungainly. I'd rather declare each tool inline e.g.
```
<Rect>
     <Tool icon={SaveIcon} onClick={...} />
     <Tool icon={ImageIcon} onClick={...} />
     ...
</Rect>
```

File: src/components/Canvas.tsx (hunk 1 above)
Comment (dai-shi) [reply src/components/Canvas.tsx:69]: Yes, the key feature of katachidraw is reusable RN component library and this allows full customization. And, it's where atoms shine.

**Comment by dai-shi** [comment]
wow, it makes me feel that i should learn a lot. thanks tons for your work.

this gives lots of insights. katachidraw is an app to show the power of jotai, so I should definitely learn and try jotai/xstate.

**Comment by dai-shi** [comment]
This helped quite a lot. I'm now starting working with jotai/xstate. I still wonder the granularity of machines, and may come back to learn from this pr. for now, closing. this.

**Pull request closed without merging**